File: Data_Structures/Hash_tables/longest_subarray.c
// Given an array of integers, find the length of the longest subarray with a sum equal to a given target value k

#include <stdio.h>

int longestSubarraySumK(int arr[], int size, int k) {
    int maxLength = 0;
    int sum = 0;
    int hash[100] = {0};

    for (int i = 0; i < size; i++) {
        sum += arr[i];

        if (sum == k) {
            maxLength = i + 1;
        }

        if (hash[sum - k] != 0) {
            if (i - hash[sum - k] + 1 > maxLength) {
                maxLength = i - hash[sum - k] + 1;
            }
        } else {
            hash[sum] = i + 1;
        }
    }

    return maxLength;
}

int main() {
    int arr[] = {3, 1, 2, -2, 4, 6, -3, 2};
    int size = sizeof(arr) / sizeof(arr[0]);
    int k = 4;

    int length = longestSubarraySumK(arr, size, k);

    printf("Length of longest subarray with sum %d: %d\n", k, length);
    printf("\n");
    
    return 0;
}

File: Data_Structures/Hash_tables/longest_subarray.py
# Given an array of integers, find the length of the longest subarray with a sum equal to a given target value k

def longest_subarray_sum_k(arr, k):
    max_length = 0
    cumulative_sum = 0
    hash_map = {0: -1}

    for i in range(len(arr)):
        cumulative_sum += arr[i]

        if cumulative_sum == k:
            max_length = i + 1

        if cumulative_sum - k in hash_map:
            length = i - hash_map[cumulative_sum - k]
            if length > max_length:
                max_length = length
        elif cumulative_sum not in hash_map:
            hash_map[cumulative_sum] = i

    return max_length

arr = [3, 1, 2, -2, 4, 6, -3, 2]
k = 4
length = longest_subarray_sum_k(arr, k)
print("Length of longest subarray with sum", k, ":", length)
